modules/generator/registry/native_histogram.go

```diff
@@ -0,0 +1,336 @@
+				_, err = appender.AppendExemplar(ref, lb.Labels(), exemplar.Exemplar{
+					Labels: convertLabelPairToLabels(bucket.Exemplar.GetLabel()),
+					Value:  bucket.Exemplar.GetValue(),
+					Ts:     timeMs,
+				})
+				if err != nil {
+					return activeSeries, err
+				}
+				bucket.Exemplar.Reset()
+			}
+		}
+
+		if !infBucketWasAdded {
+			// Add +Inf bucket
+			lb.Set(labels.BucketLabel, "+Inf")
+
+			_, err = appender.Append(0, lb.Labels(), timeMs, getIfGreaterThenZeroOr(s.histogram.GetSampleCountFloat(), s.histogram.GetSampleCount()))
+			if err != nil {
+				return activeSeries, err
+			}
+			activeSeries++
+		}
+
+		// drop "le" label again
+		lb.Del(labels.BucketLabel)
+
+		// *** Native histogram
+
+		// Decode to Prometheus representation
+		hist := promhistogram.Histogram{
+			Schema:        s.histogram.GetSchema(),
+			Count:         s.histogram.GetSampleCount(),
+			Sum:           s.histogram.GetSampleSum(),
+			ZeroThreshold: s.histogram.GetZeroThreshold(),
+			ZeroCount:     s.histogram.GetZeroCount(),
+		}
+		if len(s.histogram.PositiveSpan) > 0 {
+			hist.PositiveSpans = make([]promhistogram.Span, len(s.histogram.PositiveSpan))
+			for i, span := range s.histogram.PositiveSpan {
+				hist.PositiveSpans[i] = promhistogram.Span{
+					Offset: span.GetOffset(),
+					Length: span.GetLength(),
+				}
+			}
+		}
+		hist.PositiveBuckets = s.histogram.PositiveDelta
+		if len(s.histogram.NegativeSpan) > 0 {
+			hist.NegativeSpans = make([]promhistogram.Span, len(s.histogram.NegativeSpan))
+			for i, span := range s.histogram.NegativeSpan {
+				hist.NegativeSpans[i] = promhistogram.Span{
+					Offset: span.GetOffset(),
+					Length: span.GetLength(),
+				}
+			}
+		}
+		hist.NegativeBuckets = s.histogram.NegativeDelta
+
+		lb.Set(labels.MetricName, h.metricName)
+		_, err = appender.AppendHistogram(0, lb.Labels(), timeMs, &hist, nil)
+		if err != nil {
+			return activeSeries, err
+		}
+
+		// TODO: impact on active series from appending a histogram?
+		activeSeries += 0
+
+		if len(s.histogram.Exemplars) > 0 {
+			for _, encodedExemplar := range s.histogram.Exemplars {
+
+				// TODO are we appending the same exemplar twice?
+				e := exemplar.Exemplar{
+					Labels: convertLabelPairToLabels(encodedExemplar.Label),
+					Value:  encodedExemplar.GetValue(),
+					Ts:     convertTimestampToMs(encodedExemplar.GetTimestamp()),
+					HasTs:  true,
+				}
+				_, err = appender.AppendExemplar(0, lb.Labels(), e)
+				if err != nil {
+					return activeSeries, err
+				}
+			}
+		}
+
+	}
+
+	return
+}
+
+func (h *nativeHistogram) removeStaleSeries(staleTimeMs int64) {
+	h.seriesMtx.Lock()
+	defer h.seriesMtx.Unlock()
+	for hash, s := range h.series {
+		if s.lastUpdated.Load() < staleTimeMs {
+			delete(h.series, hash)
+			h.onRemoveSerie(h.activeSeriesPerHistogramSerie())
+		}
+	}
+}
+
+func (h *nativeHistogram) activeSeriesPerHistogramSerie() uint32 {
+	// TODO can we estimate this?
+	return 1
```

Comment:
this is interesting. i suppose this plays into how many series prom/mimir consider a native histogram?

Reply:
Right, I think we have something to learn here, since there doesn't seem to be a clean mapping between native histograms and active series.